diff -rupN Pillow-5.0.0/setup.py Pillow-5.0.0-new/setup.py
--- Pillow-5.0.0/setup.py	2018-01-01 16:03:06.000000000 +0100
+++ Pillow-5.0.0-new/setup.py	2018-01-03 11:27:55.608667037 +0100
@@ -358,8 +358,10 @@ class pil_build_ext(build_ext):
                     'aarch64': ["/usr/lib64", "/usr/lib/aarch64-linux-gnu"],
                     'arm':     ["/usr/lib/arm-linux-gnueabi"],
                     'armv71':  ["/usr/lib/arm-linux-gnueabi"],
+                    'armv7l':  ["/usr/lib"],
                     'ppc64':   ["/usr/lib64", "/usr/lib/ppc64-linux-gnu",
                                 "/usr/lib/powerpc64-linux-gnu"],
+                    'ppc64le':   ["/usr/lib64"],
                     'ppc':     ["/usr/lib/ppc-linux-gnu",
                                 "/usr/lib/powerpc-linux-gnu"],
                     's390x':   ["/usr/lib64", "/usr/lib/s390x-linux-gnu"],
