diff -rupN python-imaging-Pillow-72de37c/PIL/Image.py python-imaging-Pillow-72de37c-new/PIL/Image.py
--- python-imaging-Pillow-72de37c/PIL/Image.py	2014-04-02 00:31:55.000000000 +0200
+++ python-imaging-Pillow-72de37c-new/PIL/Image.py	2014-04-22 13:52:20.469372348 +0200
@@ -505,8 +505,11 @@ class Image:
 
     def _dump(self, file=None, format=None):
         import tempfile, os
+        suffix = ''
+        if format:
+            suffix = '.' + format
         if not file:
-            f, file = tempfile.mkstemp(format or '')
+            f, file = tempfile.mkstemp(suffix)
             os.close(f)
             
         self.load()
