diff -rupN python-imaging-Pillow-b1b88cf/Tests/test_file_webp.py python-imaging-Pillow-b1b88cf-new/Tests/test_file_webp.py
--- python-imaging-Pillow-b1b88cf/Tests/test_file_webp.py	2014-01-01 22:10:13.000000000 +0100
+++ python-imaging-Pillow-b1b88cf-new/Tests/test_file_webp.py	2014-01-03 01:14:58.873000325 +0100
@@ -28,7 +28,7 @@ def test_read_rgb():
 
     # generated with: dwebp -ppm ../../Images/lena.webp -o lena_webp_bits.ppm
     target = Image.open('Tests/images/lena_webp_bits.ppm')
-    assert_image_equal(image, target)
+    assert_image_similar(image, target, 20.0)
 
 
 def test_write_rgb():
