diff -rupN python-imaging-Pillow-b1b88cf/docs/conf.py python-imaging-Pillow-b1b88cf-new/docs/conf.py
--- python-imaging-Pillow-b1b88cf/docs/conf.py	2014-01-01 22:10:13.000000000 +0100
+++ python-imaging-Pillow-b1b88cf-new/docs/conf.py	2014-01-04 23:31:13.923990503 +0100
@@ -35,9 +35,7 @@ pygments_style = 'sphinx'
 
 ### HTML output ###
 
-from better import better_theme_path
-html_theme_path = [better_theme_path]
-html_theme = 'better'
+html_theme = 'default'
 
 html_title = "Pillow v{release} (PIL fork)".format(release=release)
 html_short_title = "Home"
